extensions/amp-analytics/0.1/amp-analytics-3p-message-queue.js

```diff
@@ -0,0 +1,187 @@
+    /** @private {boolean} */
+    this.isReady_ = false;
+
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!../../../3p/iframe-messaging-client.IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @private {string} */
+    this.messageType_ = messageType;
+
+    /** @private {!Object<string,!string|!Array<string>>} */
+    this.creativeToPendingMessages_ = {};
+
+    /** @private {function(...*)} */
+    this.throttledFlushQueue_ = throttle(this.win_, () => {
+      this.flushQueue_();
+    }, MESSAGE_THROTTLE_TIME_);
+  }
+
+  /**
+   * Indicate that a cross-domain frame is ready to receive messages, and
+   * send all messages that were previously queued for it.
+   */
+  setIsReady() {
+    this.isReady_ = true;
+    this.throttledFlushQueue_();
+  }
+
+  /**
+   * Send queued data (if there is any) to a cross-domain iframe
+   * This should never be called directly. Use this.throttledFlushQueue_()
+   * instead.
+   * @private
+   */
+  flushQueue_() {
+    if (this.isReady_ && Object.keys(this.creativeToPendingMessages_).length) {
+      const jsonMsg = /** @type {JsonObject} */ (this.buildMessage_());
+      this.iframeMessagingClient_./*OK*/sendMessage(this.messageType_, jsonMsg);
+      this.creativeToPendingMessages_ = {};
+    }
+  }
+
+  /**
+   * @return {../../../src/3p-analytics-common.AmpAnalytics3pNewCreative|
+   *          ../../../src/3p-analytics-common.AmpAnalytics3pEvent}
+   * @abstract
+   */
+  buildMessage_() {
+  }
+}
+
+export class AmpAnalytics3pNewCreativeMessageQueue extends
+  AbstractAmpAnalytics3pMessageQueue {
+
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!../../../3p/iframe-messaging-client.IframeMessagingClient}
+   *   iframeMessagingClient Facilitates cross-frame communication
+   */
+  constructor(win, iframeMessagingClient) {
+    super(win, iframeMessagingClient, AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE);
+  }
+
+  /**
+   * Enqueues an AmpAnalytics3pNewCreative message to be sent to a cross-domain
+   * iframe.
+   * @param {!string} senderId Identifies which creative is sending the message
+   * @param {string=} opt_data The data to be enqueued and then sent to the
+   * iframe
+   */
+  enqueue(senderId, opt_data) {
+    dev().assert(!this.creativeToPendingMessages_[senderId],
+        'Replacing existing extra data for: ' + senderId);
+    this.creativeToPendingMessages_[senderId] = opt_data || '';
+    this.throttledFlushQueue_();
+  }
+
+  /**
+   * Builds a message object containing the queued data
+   * @return {../../../src/3p-analytics-common.AmpAnalytics3pNewCreative}
+   * @override
+   */
+  buildMessage_() {
+    const message =
+      /** @type {../../../src/3p-analytics-common.AmpAnalytics3pNewCreative} */
+      ({
+        type: this.messageType_,
+        data: this.creativeToPendingMessages_,
+      });
+    return message;
```

Comment:
Nit: you can put the object constructor right after return, save message variable creation